File: SwiftBookingTest.Web/Models/ClientRecord.cs
using System;
using System.ComponentModel.DataAnnotations;

namespace SwiftBookingTest.Web.Models
{
    public class ClientRecord
    {
        [Key]
        public Guid? Id { get; private set; }
        public string Name { get; private set; }
        public string Address { get; private set; }
        public string Phone { get; private set; }
        public bool IsBooked { get; private set; }

        protected ClientRecord()
        {
        }

        public ClientRecord(string name, string address, string phone)
        {
            Id = Guid.NewGuid();
            Name = name;
            Address = address;
            Phone = phone;
        }

        public void Book()
        {
            IsBooked = true;
        }
    }
}

File: SwiftBookingTest.Web/Scripts/client-record.js
function ClientRecord(model) {
	var self = this

	model = model || {}

	self.name = ko.observable('')
	self.phone = ko.observable('')
	self.address = ko.observable(model.address || '')
	self.isBooked = ko.observable(model.isBooked || '')
	self.id = ko.observable(model.id || null)
	self.isBooking = ko.observable(false)

	self.canAdd = ko.computed(function() {
		return !self.id() && self.name() && self.phone() && self.address() && !self.isBooking()
	})

	self.book = function($parent) {
		if (!self.id()) {
			alert("Cannot book this record as it hasn't been saved")
			return
		}
		self.isBooking(true)
		$
			.post('/api/BookClientRecord', {
				id: self.id()
			})
			.then(
				function(response) {
					self.isBooked(true)
					$parent.bookedResponse(ko.toJSON(response, null, 2))
				},
				function(error) {
					var errorResponse = JSON.parse(error.responseText)
					
					if (!!errorResponse.code && !!errorResponse.message) {
						alert('Could not book client record: ' + errorResponse.code + ' - ' + errorResponse.message)
						$parent.bookedResponse(ko.toJSON(errorResponse, null, 2))
					} else {
						alert('Could not book client record: ' + error.statusText + ' - ' + error.responseText)
					}
				}
			)
			.always(function() {
				self.isBooking(false)
			})
		
	}

	self.toAddModel = function() {
		return {
			name: self.name(),
			phone: self.phone(),
			address: self.address(),
			id: self.id()
		}
	}
}
